src/tcp.toit

```diff
@@ -31,3 +33,84 @@ class TcpTransport implements Transport:
       with_timeout timeout:
         return Packet.deserialize reader_
     return null
+
+/**
+A transport for backing an MQTT client with TCP or TLS/TCP.
+
+Supports reconnecting to the same server.
+*/
+class ReconnectTcpTransport implements ReconnectTransport:
+  interface_   /tcp.Interface
+  use_tls_     /bool
+  certificate_ /tls.Certificate?
+  server_name_ /string?
+  root_certificates_ /List
+
+  host_ /string
+  port_ /int
+
+  transport_ /TcpTransport? := null
+  reconnecting_mutex /monitor.Mutex? := null
+
+  constructor .interface_ --host/string --port/int=1883:
+    host_ = host
+    port_ = port
+    use_tls_ = false
+    root_certificates_ = []
+    server_name_ = null
+    certificate_ = null
+
+    transport_ = new_transport_
+
+  constructor.tls .interface_ --host/string --port/int=8883
+      --root_certificates/List=[]
+      --server_name/string?=null
+      --certificate/tls.Certificate?=null:
+    host_ = host
+    port_ = port
+    use_tls_ = true
+    root_certificates_ = root_certificates
+    server_name_ = server_name
+    certificate_ = certificate
+
+    transport_ = new_transport_
+
+  send packet/Packet:
+    transport_.send packet
+
+  receive --timeout/Duration?=null -> Packet?:
+    return transport_.receive --timeout=timeout
+
+  reconnect:
+    if reconnecting_mutex:
```

Comment:
It feels a little bit odd to use a mutex for this. Maybe a latch would be more appropriate? I know you don't need a value to flow through.

An alternative would be something where you don't change the mutex all the time:

```
existing := transport_
reconnecting_mutex.do:
  transport := transport_
  if not (identical existing transport): return
  if transport: close_transport_
  transport_ = new_transport_
```

Reply:
I considered using a latch, but the comment states there should be at most one call to `get`.
It's probably true that there will only ever be two tasks that will try to reconnect, but I didn't want to rely on it.

Switched to what you proposed.